Correctly handle case-insensitive GitHub URLs (Remove duplicates)

captbaritone opened the pull request:
Depends on: #78, #79
Fixes: #74, #72, #68, #64
 
GitHub's owner and repository names are case insensitive, but we were
determining uniqueness by directly comparing strings. The result is we've
collected a bunch of duplicate data.

Unfortunately our "write-only" approach to maintaining our data means that
simply correcting our scraping will not clean up the duplicate data. Instead we
need to pro actively remove the duplicates in our tables. To that end I have
created two migrations which perform this task. They should be run in this
order:

    # Remove dupes from the `plugin_github_repos` table
    python db/migrations/dedupe_github_repo_owner.py

    # Correct our `plugin_manager_users` counts
    make build_github_index

    # Remove dupes from the `plugins` table
    python db/migrations/dedupe_plugin_repo_owner

I could not find an existing method to pro actively push the
`plugin_manager_count` from `plugin_github_repos` to the `github_bundles`
fields in `plugins`, but our scrapping cron jobs should get us back in sync
eventually.

Questions
---------

1. Are we discarding an association between vimorg and plugins?
2. Is it okay to sacrifice (and possibly reuse) the permalinks of the duplicate
   plugin entries we will delete?

Further work
------------

There are at least one other class of duplicate plugins. Plugins that were
generated from github repositories which were later renames. In this case we
update the `plugin_github_repos` table, but never proactively clean the
`plugins` table (as far as I can tell).

divad12 commented on db/migrations/dedupe_github_repo_owner.py:
```diff
@@ -0,0 +1,57 @@
+        canonical_owner_repo = (canonical['owner'], canonical['repo_name'])
+
+        if canonical_owner_repo != owner_repo:
+            print "Normalizing %s to %s for our canonical row" % (canonical_owner_repo, owner_repo)
```
nit: wrap to 79 chars per line, please?

divad12 commented on db/migrations/dedupe_github_repo_owner.py:
```diff
@@ -0,0 +1,57 @@
+        print '\nRepo with GitHub path %s occurs %s times' % (owner_repo, len(repos))
+
+        # Use the most recently scraped row as the canonical row
+        canonical = repos.pop()
```
I'm wondering what are the risks of using the most recently scraped row as the canonical row? Could there be other information or more updated info from the other rows that we may want to consolidate? For example, our count for number of dotfiles that include this plugin (`'github_bundles`).

Actually that's the only risk I can think of for now. Perhaps we can sum the `github_bundles` from all the other rows into this row?

divad12 replied:
Ah, I see you have a note in your PR description that `github_bundles` should get back in sync through the scraping cron jobs. Mind just adding a comment about that here?

divad12 commented on db/migrations/dedupe_plugin_repo_owner.py:
```diff
@@ -0,0 +1,57 @@
+        def reducer(old, new):
+            merged = db.plugins.update_plugin(old, new)
+
+            # Use the shortest slug as the canonical slug
```
This is a clever way to pick the canonical slug. :+1:

divad12 commented:
This is awesome, thanks @captbaritone.

Great migration scripts, and really thoroughly thought-out.

Re: your questions:

> Are we discarding an association between vimorg and plugins?

You're going to have to remind me here — I've forgotten a lot about how all of this works. :) What in particular is the association we might be discarding? It seems to me like you have enough of the cases covered — your `dedupe_plugin_repo_owner` migration script should preserve that association by lowercasing the `github_owner` and `github_repo_name`, no?

Other than that, I've forgotten what other places of the codebase we might be having that association.

> Is it okay to sacrifice (and possibly reuse) the permalinks of the duplicate plugin entries we will delete?

I think it's fine to sacrifice those permalinks, for now. :) This is going to make the website much more readable, and thus useable and more likely to be linked to, in the first place.

However, if we could preserve the slugs of the duplicate plugins that are deleted and check that in somewhere (maybe as a comment), then we can later add redirects with all those slugs, to do our part in slowing link rot. :)

> There are at least one other class of duplicate plugins.

Good find!

divad12 commented:
Other than that one comment, this looks good to me to merge! And once the dependencies have been merged too. :)

superscript18 commented:
Gotta say I'm super excited for this to go through.  Vim Awesome is, well, awesome, but I've been staying away from it while this duplication bug has been driving me crazy.

captbaritone commented:
Rebased and updated.

I'd like another day to double check that we aren't losing any data in the plugins table.

## Something to consider:

Unless we do something proactive about the slugs we've deleted, there's nothing stopping a future plugin from being assigned one of those slugs. A dirty hack would be to replace the list of the strings we append to slugs to make them unique. That should buy us some time to decide how to handle deleted slugs.

A _real_ solution would probably involve preserving the duplicate `plugins` rows, and just flagging them as duplicates in some way. The rest of the code would have to be updated to account for that complexity.

divad12 commented:
Great, take your time to double check and consider.

What's the precise scenario you're thinking about where a future plugin
gets assigned one of those plugins we've deleted? I'm a little unsure since
this PR resolves duplicate plugins from GitHub repo names with different
cases. Are you saying what happens if we find another plugin from a
different source (e.g. vim.org) and happen to try to assign it one of the
same slugs?

And what would the problem be in that scenario (or a different one)? That a
link from another site would be pointing to a different plugin now, or
another problem?

On Mon, Feb 29, 2016 at 10:07 PM, Jordan Eldredge <notifications@github.com>
wrote:

> Rebased and updated.
>
> I'd like another day to double check that we aren't losing any data in the
> plugins table.
> Something to consider:
>
> Unless we do something proactive about the slugs we've deleted, there's
> nothing stopping a future plugin from being assigned one of those slugs. A
> dirty hack would be to replace the list of the strings we append to slugs
> to make them unique. That should buy us some time to decide how to handle
> deleted slugs.
>
> A *real* solution would probably involve preserving the duplicate plugins
> rows, and just flagging them as duplicates in some way. The rest of the
> code would have to be updated to account for that complexity.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/divad12/vim-awesome/pull/80#issuecomment-190562104>.
>

captbaritone commented:
@divad12 Slugs are just unique identifiers, and they use the name of the plugin as a starting point. It's quite possible for plugins that are distinct to share a name just by coincidence. I believe there's even a comment in the code base which mentions that there are many people who have created plugins and called them `python.vim`.

Imagine a case like this:

1. `plugins/python-enchanted` => Is actually a dupe of `plugins/python` which lives at `github.com/some-owner/python.vim`
2. We delete `plugins/python-enchanted` and log that it should eventually redirect to `plugins/python`
3. We discover a new plugin at `github.com/another-owner/python.vim`
4. We generate a unique slug for this new plugin, and find that `plugins/python-enchanted` is unused, so we choose that.
5. We finally get around to setting up 301 redirects for deleted dupes and point `plugins/python-enchanted` to `plugins/python`
6. Poor little `github.com/another-owner/python.vim` ends up being inaccessible, since it's slug has essentially been stolen.

divad12 commented:
Gotcha, that makes sense to me. Thanks for the clear explanation.

Perhaps we can change the slug suffixes, or the way we generate other slugs
— I'm not opposed to that. :)

On Wed, Mar 2, 2016 at 10:58 AM, Jordan Eldredge <notifications@github.com>
wrote:

> @divad12 <https://github.com/divad12> Slugs are just unique identifiers,
> and they use the name of the plugin as a starting point. It's quite
> possible for plugins that are distinct to share a name just by coincidence.
> I believe there's even a comment in the code base which mentions that there
> are many people who have created plugins and called them python.vim.
>
> Imagine a case like this:
>
>    1. plugins/python-enchanted => Is actually a dupe of plugins/python
>    which lives at github.com/some-owner/python.vim
>    2. We delete plugins/python-enchanted and log that it should
>    eventually redirect to plugins/python
>    3. We discover a new plugin at github.com/another-owner/python.vim
>    4. We generate a unique slug for this new plugin, and find that
>    plugins/python-enchanted is unused, so we choose that.
>    5. We finally get around to setting up 301 redirects for deleted dupes
>    and point plugins/python-enchanted to plugins/python
>    6. Poor little github.com/another-owner/python.vim ends up being
>    inaccessible, since it's slug has essentially been stolen.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/divad12/vim-awesome/pull/80#issuecomment-191372321>.
>

captbaritone commented:
Looked through things a bit more, and I think we should do some extra effort to make sure user supplied information like category and tags are preserved. When we merge plugins, we should probably merge tags and keep a category if any of the dupes had one.

Getting there!

divad12 commented:
:+1: that sounds good!

captbaritone commented:
Alright, I think this should be ready to go. It now preserves user-supplied info and has a test to assert that the merging is being done correctly.

Review: @jonafato 
Pythonic review, if he's intersted: @tulioz

tulioz commented on db/migrations/dedupe_plugin_repo_owner.py:
```diff
@@ -0,0 +1,87 @@
+            new['category'] = old['category']
+
+        # Merge tags
+        new['tags'] = list(set(new['tags'] + old['tags']))
```
```set()``` returns a list, no need to wrap it.

jonafato replied:
`set` is a type in Python. It's possible (probable?) that rethinkdb will automatically convert sets to lists, but the conversion being done here is probably safer, and it keeps types consistent with other parts of the code.

tulioz replied:
I was thinking sorted, that's what I get for being tired and code reviewing at the same time. Thanks for the correction.

tulioz commented on db/migrations/dedupe_plugin_repo_owner.py:
```diff
@@ -0,0 +1,87 @@
+
+        # Use the plugin with the shortest slug as the new plugin
+        if len(old['slug']) < len(new['slug']):
+            new, old = (old, new)
```
No need to convert old/new into a tuple first.

```python
new, old = old, new
```

Will also work.

jonafato replied:
:+1: The parentheses aren't needed. (The [comma](https://wiki.python.org/moin/TupleSyntax#Multiple_Element_Tuples) creates the `tuple`, so the parentheses are simply stylistic here.)

tulioz commented:
:+1: the Python looks good to me, the comments I left are trivial.

jonafato commented:
Reviewing this now. It looks like there's a bit of history here (with the related pull requests and issues), so it may take me a bit of time to catch up.

captbaritone commented:
@jonafato yeah. Lots to tease apart. A full review would be appreciated if you have the time. If not, a cursory sanity check would be great.

Feel free to ask for clarification if you have questions.

I would love to get this out by Tuesday if possible.

jonafato commented on db/migrations/dedupe_github_repo_owner.py:
```diff
@@ -0,0 +1,61 @@
+        # Use the most recently scraped row as the canonical row
+        canonical = repos.pop()
+
+        assert canonical
```
What's the purpose of this `assert` statement here? As written, if `canonical` is falsey in an iteration of the loop, no later iterations will be run. Would this be better as something like the following?

```python
if not canonical:
    continue
````

captbaritone replied:
It's intended as a sanity check. Basically: "if canonical is falsey here, something has gone wrong and we should just crash out"

It also doubles as a sort of comment: "from this point on, we're operating under the assumption that we have a canonical item"

--
Jordan Eldredge

Please excuse any typos or terseness; this email was sent from a mobile device. 

> On Mar 5, 2016, at 3:20 PM, Jon Banafato <notifications@github.com> wrote:
> 
> In db/migrations/dedupe_github_repo_owner.py:
> 
> > +
> > +    # Get the most recently scraped row in each group first
> > +    query = query.order_by(r.desc('last_scaped_at'))
> > +
> > +    grouped_repos = query.run(r_conn())
> > +
> > +    for owner_repo, repos in grouped_repos.iteritems():
> > +
> > +        print '\nRepo with GitHub path %s occurs %s times' % (
> > +                owner_repo,
> > +                len(repos))
> > +
> > +        # Use the most recently scraped row as the canonical row
> > +        canonical = repos.pop()
> > +
> > +        assert canonical
> What's the purpose of this assert statement here? As written, if canonical is falsey in an iteration of the loop, no later iterations will be run. Would this be better as something like the following?
> 
> if not canonical:
>     continue
> —
> Reply to this email directly or view it on GitHub.
>

jonafato replied:
This would potentially crash out after data has been modified in the database, though, correct? i.e. calling `run` in iterations prior to one during which an `AssertionError` is raised will have migrated some data (but not all that was intended). If this is the intended execution plan, there's nothing inherently wrong with it. My personal approach would be either to check for errors and bail out before anything gets modified or to migrate everything that can be and log the problematic records for further review. Just food for thought.

captbaritone replied:
Good point. I had honestly been using this as part of my debug process while building the script. In that case, I was working with a database dump which I could easily restore. Obviously I'll make a backup of our production DB before running this script (we make one every day anyway), so I could always restore from that if needed.

I'll leave it, since I'm 99.99% sure it could never be falsy anyway (given the way the rethinkdb api works).

Good things to consider though. I'm still formulating for myself what the proper way/time to use `assert` actually is.

jonafato commented on db/migrations/dedupe_github_repo_owner.py:
```diff
@@ -0,0 +1,61 @@
+                len(repos))
+
+        # Use the most recently scraped row as the canonical row
+        canonical = repos.pop()
```
I suspect this is not the case, but I'll ask anyway so someone more familiar with the data model can make the correct call: are there valid instances where `repos` is an empty list? If so, this will raise an `IndexError`, which should be handled.

captbaritone replied:
The rethinkdb  `group` method insures this will always be non-empty.

jonafato replied:
:+1: Thanks for the clarification.

jonafato commented on db/migrations/dedupe_plugin_repo_owner.py:
```diff
@@ -0,0 +1,87 @@
+        db.plugins.insert(canonical, conflict='replace')
+        updated += 1
+
+        dupes = [dupe for dupe in plugins if dupe['slug'] is not canonical['slug']]
```
```python
dupe['slug'] is not canonical['slug']
```
should probably be
```python
dupe['slug'] != canonical['slug']
```

jonafato commented on db/plugins.py:
```diff
@@ -170,6 +170,8 @@ def insert(plugins, *args, **kwargs):
             plugin['normalized_name'] = _normalize_name(plugin)
 
+        assert plugin['github_owner'] == plugin['github_owner'].lower()
+        assert plugin['github_repo_name'] == plugin['github_repo_name'].lower()
```
Raising a `ValueError` would be more appropriate here.

jonafato commented:
Should [these lines](https://github.com/divad12/vim-awesome/blob/c02cd9812014cbe3cc724be3483efab4545f29a6/tools/scrape/github.py#L273-L274) be updated to be lowercased as well?

jonafato commented:
Most of my comments are nit-picks or stylistic and won't affect the critical path. Great work, looks like this is a much needed fix!

captbaritone commented:
@jonafato Thanks for the excellent review. A number of great catches. I've opted to move the owner/repo normalization for plugins inside of `db.plugins.insert`. This handles the missed case that you pointed out, and a few others. It's probably overkill in some places, but I figure it's better to be safe than sorry.

I've added a few additional commits, and I would appreciate it if you have time for a second look. Once everything looks good, I'll squash this down to one commit and merge.

jonafato commented:
New commits look great! +1 on squashing commits (though it's almost a shame to lose such well formed and descriptive individual commit messages).

captbaritone commented:
squashed
